sentry-android-core/src/main/java/io/sentry/android/core/internal/util/SentryFrameMetricsCollector.java

```diff
@@ -83,18 +89,83 @@ public SentryFrameMetricsCollector(
     // start a profile, we wouldn't have the current activity and couldn't get the frameMetrics.
     ((Application) context).registerActivityLifecycleCallbacks(this);
 
+    // Most considerations regarding timestamps of frames are inspired from JankStats library:
+    // https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:metrics/metrics-performance/src/main/java/androidx/metrics/performance/JankStatsApi24Impl.kt
+
+    // The Choreographer instance must be accessed on the main thread
+    new Handler(Looper.getMainLooper()).post(() -> choreographer = Choreographer.getInstance());
+    // Let's get the last frame timestamp from the choreographer private field
+    try {
+      choreographerLastFrameTimeField = Choreographer.class.getDeclaredField("mLastFrameTimeNanos");
+      choreographerLastFrameTimeField.setAccessible(true);
+    } catch (NoSuchFieldException e) {
+      options
+          .getLogger()
+          .log(SentryLevel.ERROR, "Unable to get the frame timestamp from the choreographer: ", e);
+    }
     frameMetricsAvailableListener =
         (window, frameMetrics, dropCountSinceLastInvocation) -> {
-          float refreshRate =
+          final long now = System.nanoTime();
+          final float refreshRate =
               buildInfoProvider.getSdkInfoVersion() >= Build.VERSION_CODES.R
                   ? window.getContext().getDisplay().getRefreshRate()
                   : window.getWindowManager().getDefaultDisplay().getRefreshRate();
+
+          final long cpuDuration = getFrameCpuDuration(frameMetrics);
+          long startTime = getFrameStartTimestamp();
+          // If we couldn't get the timestamp through reflection, we use current time
+          if (startTime == -1) {
+            startTime = now - cpuDuration;
+          }
+          // Let's "adjust" the start time of a frame to be after the end of the previous frame
+          startTime = Math.max(startTime, lastFrameEndNanos);
+          // Let's avoid emitting duplicates (start time equals to last start time)
+          if (startTime == lastFrameStartNanos) {
+            return;
+          }
+          lastFrameStartNanos = startTime;
+          lastFrameEndNanos = startTime + cpuDuration;
+
           for (FrameMetricsCollectorListener l : listenerMap.values()) {
-            l.onFrameMetricCollected(frameMetrics, refreshRate);
+            l.onFrameMetricCollected(lastFrameEndNanos, cpuDuration, refreshRate);
           }
         };
   }
 
+  /**
+   * Return the internal timestamp in the choreographer of the last frame start timestamp through
+   * reflection.
+   */
+  private long getFrameStartTimestamp() {
```

Comment:
`m`: I've also seen JankStats uses a different implementation to get the frame start time from [API 26 and above](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:metrics/metrics-performance/src/main/java/androidx/metrics/performance/JankStatsApi26Impl.kt).

Should we consider this in our implementation as well? Not sure if it's more accurate, maybe needs some testing (or maybe it just avoids reflection on new Android versions, which is also good for us?).